src/net/tcp/stream.rs

```diff
@@ -113,14 +119,13 @@ impl TcpStream {
     ///             let mut n = 0;
     ///             let mut buf = vec![0u8; 4096];
     ///             loop {
-    ///                 let (result, nbuf) = stream.read(buf).await;
-    ///                 buf = nbuf;
+    ///                 let (result, slice) = stream.read(buf).await;
     ///                 let read = result.unwrap();
     ///                 if read == 0 {
     ///                     break;
     ///                 }
     ///
-    ///                 let (res, slice) = stream.write_all(buf.slice(..read)).await;
+    ///                 let (res, slice) = stream.write_all(slice.slice(..read)).await;
```

Comment:
Wonder how many times people will forget they need to qualify the slice they were given by the value returned. 

And will we come up with a macro, kind of like ready!, for reads that return a result and the slice where the error is returned if the result was an error, and otherwise unwraps the read integer?

Not in this PR of course.

Reply:
I've been thinking of giving the returned slice an additional purpose. Perhaps the slice returned by `read` could be truncated to the bytes that were read.

Reply:
@mzabaluev I absolutely agress with this - following a read, the buffer you get back should only contain valid data.

Reply:
I think that is current behavior?

Reply:
@ollie-etl No, both in the current API and in this PR the returned buffer/slice has the same range as the input. The length of the bytes updated by read is returned in `Ok(n)`, but the buffer view may also expose pre-existing data in the rest of its initialized part.

Reply:
Resolving this. I think we agree we're getting as close a behavior as possible to before the PR, the Slice<T> being returned not withstanding.